File: ts/scripts/build-definitions.ts
import * as fs from "node:fs";
import * as path from "node:path";
import * as os from "node:os";
import {execSync} from "node:child_process";
import {generate} from "ts-to-zod";
import {Project} from "ts-morph";

const args = process.argv.slice(2);
const versionIndex = args.indexOf("--version");
if (versionIndex === -1 || !args[versionIndex + 1]) {
    console.error("Usage: npm run build:definitions -- --version <version>");
    process.exit(1);
}
const version = args[versionIndex + 1];

const REPO = "code0-tech/code0-definition";
const OUTPUT_DIR = path.resolve("src/definitions");

// ── Helpers ───────────────────────────────────────────────────────────────────

function toPascalCase(str: string): string {
    return str
        .replace(/::/g, "_")
        .split(/[_-]/)
        .map(p => p.charAt(0).toUpperCase() + p.slice(1).toLowerCase())
        .join("");
}

function fileNameToClassName(file: string): string {
    return toPascalCase(path.basename(file, ".proto.json"));
}

function toSchemaName(identifier: string): string {
    // HTTP_METHOD → httpMethodSchema
    return identifier.toLowerCase().replace(/_([a-z])/g, (_, c) => c.toUpperCase()) + "Schema";
}

function srcRelativePath(relModule: string, ...segments: string[]): string {
    const depth = relModule.split(path.sep).filter(Boolean).length + 1;
    return "../".repeat(depth) + segments.join("/");
}

type Translation = { code: string; content: string };

function renderTranslations(decorator: string, arr: Translation[]): string[] {
    if (!arr?.length) return [];
    return [
        `@${decorator}(`,
        ...arr.map(t => `    {code: "${t.code}", content: ${JSON.stringify(t.content)}},`),
        `)`,
    ];
}

function renderTranslationsInline(arr: Translation[]): string {
    return arr.map(t => `{code: "${t.code}", content: ${JSON.stringify(t.content)}}`).join(", ");
}

// ── Data type definitions ─────────────────────────────────────────────────────

interface DataTypeDef {
    identifier: string;
    type: string;
    genericKeys: string[];
    relModule: string;
    className: string;
    fileName: string;
}

function buildCombinedSource(defs: DataTypeDef[]): string {
    return defs.map(({identifier, type, genericKeys}) => {
        let resolved = type;
        if (/\[.*in keyof/.test(resolved)) {
            // Identity mapped type (e.g. { [K in keyof T]: T[K] }) → Record<string, unknown>
            resolved = "Record<string, unknown>";
        } else {
            for (const key of genericKeys) {
                resolved = resolved.replace(new RegExp(`\\b${key}\\b`, "g"), "unknown");
            }
        }
        return `export type ${identifier} = ${resolved};`;
    }).join("\n");
}

function relativeImport(fromRelModule: string, toDef: DataTypeDef, toFile: string): string {
    const from = fromRelModule.split(path.sep).filter(Boolean);
    const to = toDef.relModule.split(path.sep).filter(Boolean);
    let common = 0;
    while (common < from.length && common < to.length && from[common] === to[common])
        common++;
    const up = "../".repeat(from.length - common);
    const down = [...to.slice(common), toFile].join("/");
    return up ? up + down : "./" + down;
}

// ── Meta-decorator helpers ────────────────────────────────────────────────────

function collectMetaDecorators(json: Record<string, unknown>): string[] {
    return [
        "Identifier",
        ...(json.signature ? ["Signature"] : []),
        ...(Array.isArray(json.name) && json.name.length ? ["Name"] : []),
        ...(Array.isArray(json.description) && json.description.length ? ["Description"] : []),
        ...(Array.isArray(json.documentation) && json.documentation.length ? ["Documentation"] : []),
        ...(Array.isArray(json.displayMessage) && json.displayMessage.length ? ["DisplayMessage"] : []),
        ...(Array.isArray(json.alias) && json.alias.length ? ["Alias"] : []),
        ...(json.displayIcon ? ["DisplayIcon"] : []),
    ];
}

function buildMetaDecorators(json: Record<string, unknown>, identifierKey: string): string[] {
    return [
        `@Identifier(${JSON.stringify(json[identifierKey])})`,
        ...(json.signature
            ? [`@Signature(${JSON.stringify(json.signature)})`]
            : []),
        ...(Array.isArray(json.name) && json.name.length
            ? renderTranslations("Name", json.name as Translation[])
            : []),
        ...(Array.isArray(json.description) && json.description.length
            ? renderTranslations("Description", json.description as Translation[])
            : []),
        ...(Array.isArray(json.documentation) && json.documentation.length
            ? renderTranslations("Documentation", json.documentation as Translation[])
            : []),
        ...(Array.isArray(json.displayMessage) && json.displayMessage.length
            ? renderTranslations("DisplayMessage", json.displayMessage as Translation[])
            : []),
        ...(Array.isArray(json.alias) && json.alias.length
            ? renderTranslations("Alias", json.alias as Translation[])
            : []),
        ...(json.displayIcon
            ? [`@DisplayIcon(${JSON.stringify(json.displayIcon)})`]
            : []),
    ];
}

// ── Generators ────────────────────────────────────────────────────────────────

function buildSettingDecorators(settings: any[]): string[] {
    return settings.flatMap(s => {
        const props = [
            `identifier: ${JSON.stringify(s.identifier)},`,
            ...(s.unique && s.unique !== "NONE" ? [`unique: ${JSON.stringify(s.unique)} as any,`] : []),
            ...(Array.isArray(s.name) && s.name.length ? [`name: [${renderTranslationsInline(s.name)}],`] : []),
            ...(Array.isArray(s.description) && s.description.length ? [`description: [${renderTranslationsInline(s.description)}],`] : []),
            ...(s.optional ? [`optional: true,`] : []),
            ...(s.hidden ? [`hidden: true,`] : []),
        ];
        return [`@RuntimeEventSetting({`, ...props.map(p => `    ${p}`), `})`];
    });
}

function buildParameterDecorators(params: any[]): string[] {
    return params.flatMap(p => {
        const props = [
            `runtimeName: ${JSON.stringify(p.runtimeName)},`,
            ...(Array.isArray(p.name) && p.name.length ? [`name: [${renderTranslationsInline(p.name)}],`] : []),
            ...(Array.isArray(p.description) && p.description.length ? [`description: [${renderTranslationsInline(p.description)}],`] : []),
            ...(p.optional ? [`optional: true,`] : []),
            ...(p.hidden ? [`hidden: true,`] : []),
        ];
        return [`@RuntimeParameter({`, ...props.map(p => `    ${p}`), `})`];
    });
}

function generateRuntimeFlowType(json: Record<string, unknown>, className: string, relModule: string): string {
    const src = (seg: string) => srcRelativePath(relModule, seg);
    const linkedIds = json.linkedDataTypeIdentifiers as string[] | undefined;
    const settings = json.runtimeSettings as any[] | undefined;
    const metaDecorators = collectMetaDecorators(json);

    const imports = [
        `import {RuntimeEventRunnable} from "${src("models/runtime_event.model")}";`,
        `import {${metaDecorators.join(", ")}} from "${src("decorators/meta")}";`,
        ...(linkedIds?.length ? [`import {LinkedDataTypeIdentifiers} from "${src("decorators/function")}";`] : []),
        ...(settings?.length ? [`import {RuntimeEventSetting} from "${src("decorators/runtime-event")}";`] : []),
    ];

    const decorators = [
        ...buildMetaDecorators(json, "identifier"),
        ...(linkedIds?.length ? [`@LinkedDataTypeIdentifiers(${linkedIds.map(s => JSON.stringify(s)).join(", ")})`] : []),
        ...buildSettingDecorators(settings ?? []),
    ];

    return [
        ...imports,
        "",
        ...decorators,
        `export class ${className} implements RuntimeEventRunnable {}`,
        "",
    ].join("\n");
}

function generateRuntimeFunction(json: Record<string, unknown>, className: string, relModule: string): string {
    const src = (seg: string) => srcRelativePath(relModule, seg);
    const linkedIds = json.linkedDataTypeIdentifiers as string[] | undefined;
    const params = json.runtimeParameterDefinitions as any[] | undefined;
    const metaDecorators = collectMetaDecorators(json);
    const funcDecorators = [
        "OmitFunctionDefinition",
        ...(json.throwsError ? ["ThrowsError"] : []),
        ...(params?.length ? ["RuntimeParameter"] : []),
        ...(linkedIds?.length ? ["LinkedDataTypeIdentifiers"] : []),
    ];

    const imports = [
        `import {RuntimeFunctionRunnable} from "${src("models/runtime_function.model")}";`,
        `import {${metaDecorators.join(", ")}} from "${src("decorators/meta")}";`,
        `import {${funcDecorators.join(", ")}} from "${src("decorators/function")}";`,
        `import type {PlainValue} from "@code0-tech/tucana/helpers";`,
    ];

    const decorators = [
        `@OmitFunctionDefinition()`,
        ...buildMetaDecorators(json, "runtimeName"),
        ...(json.throwsError ? [`@ThrowsError()`] : []),
        ...(linkedIds?.length ? [`@LinkedDataTypeIdentifiers(${linkedIds.map(s => JSON.stringify(s)).join(", ")})`] : []),
        ...buildParameterDecorators(params ?? []),
    ];

    return [
        ...imports,
        "",
        ...decorators,
        `export class ${className} implements RuntimeFunctionRunnable {`,
        `    run(..._args: (PlainValue | undefined)[]): Promise<PlainValue> | PlainValue {`,
        `        throw new Error(\`${className}.run() is not implemented\`);`,
        `    }`,
        `}`,
        "",
    ].join("\n");
}

// ── Directory walker ──────────────────────────────────────────────────────────

type DefHandler = (
    json: Record<string, unknown>,
    className: string,
    fileName: string,
    relModule: string,
    typeFolder: string,
) => void;

function walkDefs(dir: string, relModule: string, handler: DefHandler): void {
    for (const entry of fs.readdirSync(dir, {withFileTypes: true})) {
        const fullPath = path.join(dir, entry.name);
        if (entry.isDirectory()) {
            walkDefs(fullPath, path.join(relModule, entry.name), handler);
            continue;
        }
        if (!entry.name.endsWith(".proto.json")) continue;
        const json = JSON.parse(fs.readFileSync(fullPath, "utf-8")) as Record<string, unknown>;
        const typeFolder = relModule.split(path.sep).at(-1) ?? "";
        const className = fileNameToClassName(entry.name);
        const fileName = path.basename(entry.name, ".proto.json");
        handler(json, className, fileName, relModule, typeFolder);
    }
}

// ── Main ──────────────────────────────────────────────────────────────────────

async function main() {
    const url = `https://github.com/${REPO}/releases/download/${version}/definitions.zip`;
    console.log(`Downloading ${url}...`);

    const response = await fetch(url);
    if (!response.ok) throw new Error(`Download failed: ${response.status} ${response.statusText}`);

    const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), "hercules-definitions-"));
    const zipPath = path.join(tmpDir, "definitions.zip");
    const extractDir = path.join(tmpDir, "extracted");

    fs.writeFileSync(zipPath, Buffer.from(await response.arrayBuffer()));
    fs.mkdirSync(extractDir, {recursive: true});
    execSync(`unzip -q "${zipPath}" -d "${extractDir}"`);

    const defsDir = path.join(extractDir, "definitions");

    // ── Collect all data type definitions ─────────────────────────────────────
    const dataTypeDefs: DataTypeDef[] = [];
    walkDefs(defsDir, "", (json, className, fileName, relModule, typeFolder) => {
        if (typeFolder !== "data_types" || !json.identifier || !json.type) return;
        dataTypeDefs.push({
            identifier: json.identifier as string,
            type: json.type as string,
            genericKeys: (json.genericKeys as string[]) ?? [],
            relModule,
            className,
            fileName,
        });
    });

    // ── Generate all schemas via ts-to-zod ────────────────────────────────────
    const combinedSource = buildCombinedSource(dataTypeDefs);
    const {getZodSchemasFile, errors} = generate({sourceText: combinedSource});
    if (errors.length) console.warn("ts-to-zod warnings:", errors);
    const schemasSource = getZodSchemasFile("schemas");

    // ── Parse ts-to-zod output with ts-morph ──────────────────────────────────
    const morphProject = new Project({useInMemoryFileSystem: true});
    const sf = morphProject.createSourceFile("schemas", schemasSource);

    const schemaBodyMap = new Map<string, string>();
    for (const stmt of sf.getVariableStatements()) {
        for (const decl of stmt.getDeclarations()) {
            const init = decl.getInitializer()?.getText();
            if (init) schemaBodyMap.set(decl.getName(), init);
        }
    }

    const schemaToFile = new Map<string, DataTypeDef>();
    for (const def of dataTypeDefs) {
        schemaToFile.set(toSchemaName(def.identifier), def);
    }

    // ── Write output ──────────────────────────────────────────────────────────
    fs.rmSync(OUTPUT_DIR, {recursive: true, force: true});
    fs.mkdirSync(OUTPUT_DIR, {recursive: true});

    let count = 0;
    const allSchemaNames = new Set(schemaBodyMap.keys());
    const generatedFiles: string[] = [];

    for (const def of dataTypeDefs) {
        const schemaName = toSchemaName(def.identifier);
        const body = schemaBodyMap.get(schemaName);
        if (!body) continue;

        const refs = [
            ...new Set(
                [...body.matchAll(/\b(\w+Schema)\b/g)]
                    .map(m => m[1])
                    .filter(n => n !== schemaName && allSchemaNames.has(n))
            ),
        ];
        const typeParams = def.genericKeys.length > 0
            ? `<${def.genericKeys.map(k => `${k} = unknown`).join(", ")}>`
            : "";

        const lines = [
            `import {z} from "zod";`,
            ...refs.flatMap(ref => {
                const refDef = schemaToFile.get(ref);
                return refDef ? [`import {${ref}} from "${relativeImport(def.relModule, refDef, refDef.fileName)}";`] : [];
            }),
            ``,
            `export const ${schemaName} = ${body};`,
            `export type ${def.className}${typeParams} = z.infer<typeof ${schemaName}>;`,
            ``,
        ];

        const outDir = path.join(OUTPUT_DIR, def.relModule);
        fs.mkdirSync(outDir, {recursive: true});
        fs.writeFileSync(path.join(outDir, `${def.fileName}.ts`), lines.join("\n"));
        generatedFiles.push([def.relModule, def.fileName].filter(Boolean).join("/"));
        count++;
    }

    walkDefs(defsDir, "", (json, className, fileName, relModule, typeFolder) => {
        let content: string | null = null;
        if (typeFolder === "runtime_flow_types") content = generateRuntimeFlowType(json, className, relModule);
        else if (typeFolder === "runtime_functions") content = generateRuntimeFunction(json, className, relModule);
        if (content === null) return;

        const outDir = path.join(OUTPUT_DIR, relModule);
        fs.mkdirSync(outDir, {recursive: true});
        fs.writeFileSync(path.join(outDir, `${fileName}.ts`), content);
        generatedFiles.push([relModule, fileName].filter(Boolean).join("/"));
        count++;
    });

    const indexContent = generatedFiles
        .sort()
        .map(f => `export * from "./${f}";`)
        .join("\n") + "\n";
    fs.writeFileSync(path.join(OUTPUT_DIR, "index.ts"), indexContent);

    fs.rmSync(tmpDir, {recursive: true, force: true});
    console.log(`Generated ${count} definition files in src/definitions/`);
}

main().catch(err => { console.error(err); process.exit(1); });

File: py/scripts/build_definitions.py
"""Generate the ``hercules/definitions`` package from a code0-definition release.

Python port of ``ts/scripts/build-definitions.ts``. Downloads the upstream
``definitions.zip``, then for every definition emits a Python module:

* data types      -> a schema variable + ``register_schema`` call
* runtime funcs   -> a decorated ``RuntimeFunctionRunnable`` subclass
* runtime flows   -> a decorated ``RuntimeEventRunnable`` subclass

Usage::

    python scripts/build_definitions.py --version def-0.0.35
"""
from __future__ import annotations

import argparse
import io
import json
import os
import re
import shutil
import sys
import urllib.request
import zipfile
from pathlib import Path
from typing import Dict, List, Optional, Set, Tuple

# ``build_definitions`` runs standalone; make sure the package is importable.
sys.path.insert(0, str(Path(__file__).resolve().parent.parent))
from hercules.internal.pydantic_codegen import ts_to_pydantic

REPO = "code0-tech/code0-definition"
OUTPUT_DIR = Path(__file__).resolve().parent.parent / "hercules" / "definitions"


# ── Naming helpers ─────────────────────────────────────────────────────────────

def to_pascal_case(value: str) -> str:
    value = value.replace("::", "_")
    parts = [p for p in _split_words(value) if p]
    return "".join(p[:1].upper() + p[1:].lower() for p in parts)


def _split_words(value: str) -> List[str]:
    out, cur = [], ""
    for ch in value:
        if ch in "_-":
            out.append(cur)
            cur = ""
        else:
            cur += ch
    out.append(cur)
    return out


def sanitize_segment(segment: str) -> str:
    """Make a directory name a valid Python package name (``rest-action`` -> ``rest_action``)."""
    return segment.replace("-", "_")


def to_model_name(identifier: str) -> str:
    """Class name for a data type's generated Pydantic model (``OBJECT`` -> ``Object``)."""
    name = to_pascal_case(identifier)
    if not name or not name[0].isalpha():
        name = "DataType" + name
    return name


_TS_IDENT_RE = re.compile(r"^[A-Za-z_$][A-Za-z0-9_$]*$")


def module_dotted(rel_module: str, file_name: str) -> str:
    segments = [sanitize_segment(s) for s in Path(rel_module).parts]
    return ".".join(["hercules", "definitions", *segments, file_name])


# ── Definition records ─────────────────────────────────────────────────────────

class DataTypeDef:
    def __init__(self, identifier, type_str, generic_keys, rel_module, file_name):
        self.identifier = identifier
        self.type_str = type_str
        self.generic_keys = generic_keys
        self.rel_module = rel_module
        self.file_name = file_name


# ── Translation / decorator rendering ──────────────────────────────────────────

def _py_translations(items) -> str:
    return "[" + ", ".join(
        f'{{"code": "{t["code"]}", "content": {json.dumps(t["content"])}}}' for t in items
    ) + "]"


def _translation_decorator(name: str, items) -> List[str]:
    if not items:
        return []
    args = ", ".join(
        f'{{"code": "{t["code"]}", "content": {json.dumps(t["content"])}}}' for t in items
    )
    return [f"@{name}({args})"]


def _meta_decorators(json_obj: dict, identifier_key: str) -> Tuple[List[str], List[str]]:
    """Return (decorator lines, imported decorator names)."""
    lines: List[str] = [f"@Identifier({json.dumps(json_obj[identifier_key])})"]
    names = {"Identifier"}
    if json_obj.get("signature"):
        lines.append(f"@Signature({json.dumps(json_obj['signature'])})")
        names.add("Signature")
    for key, deco in (
        ("name", "Name"),
        ("description", "Description"),
        ("documentation", "Documentation"),
        ("displayMessage", "DisplayMessage"),
        ("alias", "Alias"),
    ):
        if json_obj.get(key):
            lines += _translation_decorator(deco, json_obj[key])
            names.add(deco)
    if json_obj.get("displayIcon"):
        lines.append(f"@DisplayIcon({json.dumps(json_obj['displayIcon'])})")
        names.add("DisplayIcon")
    return lines, sorted(names)


def _parameter_decorators(params) -> List[str]:
    out = []
    for p in params or []:
        props = [f'"runtime_name": {json.dumps(p["runtimeName"])}']
        if p.get("name"):
            props.append(f'"name": {_py_translations(p["name"])}')
        if p.get("description"):
            props.append(f'"description": {_py_translations(p["description"])}')
        if p.get("optional"):
            props.append('"optional": True')
        if p.get("hidden"):
            props.append('"hidden": True')
        out.append("@Parameter({" + ", ".join(props) + "})")
    return out


def _setting_decorators(settings) -> List[str]:
    out = []
    for s in settings or []:
        props = [f'"identifier": {json.dumps(s["identifier"])}']
        if s.get("unique") and s["unique"] != "NONE":
            props.append(f'"unique": {json.dumps(s["unique"])}')
        if s.get("name"):
            props.append(f'"name": {_py_translations(s["name"])}')
        if s.get("description"):
            props.append(f'"description": {_py_translations(s["description"])}')
        if s.get("optional"):
            props.append('"optional": True')
        if s.get("hidden"):
            props.append('"hidden": True')
        out.append("@EventSetting({" + ", ".join(props) + "})")
    return out


# ── File generators ────────────────────────────────────────────────────────────

def generate_runtime_function(json_obj: dict, class_name: str) -> str:
    params = json_obj.get("runtimeParameterDefinitions")
    meta_lines, meta_names = _meta_decorators(json_obj, "runtimeName")
    func_names = []
    if json_obj.get("throwsError"):
        func_names.append("ThrowsError")
    if params:
        func_names.append("Parameter")

    lines = [
        "from hercules.models.runtime_function import RuntimeFunctionRunnable",
        f"from hercules.decorators.meta import {', '.join(meta_names)}",
        "from hercules.decorators.runtime_function import OmitRuntimeFunction",
    ]
    if func_names:
        lines.append(f"from hercules.decorators.function import {', '.join(func_names)}")
    lines += ["", "", "@OmitRuntimeFunction()"]
    lines += meta_lines
    if json_obj.get("throwsError"):
        lines.append("@ThrowsError()")
    lines += _parameter_decorators(params)
    lines += [
        f"class {class_name}(RuntimeFunctionRunnable):",
        "    def run(self, *args):",
        f'        raise NotImplementedError("{class_name}.run() is not implemented")',
        "",
    ]
    return "\n".join(lines)


def generate_runtime_flow_type(json_obj: dict, class_name: str) -> str:
    settings = json_obj.get("runtimeSettings")
    meta_lines, meta_names = _meta_decorators(json_obj, "identifier")

    lines = [
        "from hercules.models.runtime_event import RuntimeEventRunnable",
        f"from hercules.decorators.meta import {', '.join(meta_names)}",
    ]
    if settings:
        lines.append("from hercules.decorators.event import EventSetting")
    lines += ["", ""]
    lines += meta_lines
    lines += _setting_decorators(settings)
    lines += [f"class {class_name}(RuntimeEventRunnable):", "    pass", ""]
    return "\n".join(lines)


def _strip_codegen_header(code: str) -> str:
    """Drop the leading ``# generated by datamodel-codegen`` banner."""
    lines = code.splitlines()
    start = 0
    while start < len(lines) and (
        lines[start].startswith("#") or not lines[start].strip()
    ):
        start += 1
    return "\n".join(lines[start:]).strip()


def _reference_preamble(dt: "DataTypeDef", ref_map, generic_key_set: Set[str]) -> str:
    """Declare every other data-type identifier / generic key as an opaque TS type.

    The referenced data types are only *registered* for reference resolution, so it
    is enough for the generator to see them as opaque (``unknown``) here; the exact
    cross-references are preserved separately in the emitted ``TypeString``.
    """
    names = set(ref_map) | set(generic_key_set) | set(dt.generic_keys)
    names.discard(dt.identifier)
    stubs = [
        f"type {name} = unknown;"
        for name in sorted(names)
        if _TS_IDENT_RE.match(name)
    ]
    return "\n".join(stubs)


def generate_data_type(dt: "DataTypeDef", ref_map, generic_key_set) -> str:
    class_name = to_model_name(dt.identifier)
    preamble = _reference_preamble(dt, ref_map, generic_key_set)

    try:
        model_code = _strip_codegen_header(
            ts_to_pydantic(dt.type_str, class_name, preamble=preamble)
        )
    except Exception as exc:  # noqa: BLE001 - keep the build resilient per-definition
        print(f"  ! {dt.identifier}: falling back to RootModel[Any] ({exc})")
        model_code = (
            "from typing import Any\n"
            "from pydantic import RootModel\n\n\n"
            f"class {class_name}(RootModel[Any]):\n    pass"
        )

    lines = [
        model_code,
        "",
        "",
        "from hercules.internal.schema import register_schema",
        f"register_schema({class_name}, {json.dumps(dt.identifier)})",
        "",
    ]
    return "\n".join(lines)


# ── Directory walking ──────────────────────────────────────────────────────────

def walk_defs(defs_dir: Path):
    for path in sorted(defs_dir.rglob("*.json")):
        if path.name == "module.json":
            continue
        rel_module = str(path.parent.relative_to(defs_dir))
        type_folder = path.parent.name
        class_name = to_pascal_case(path.stem)
        file_name = path.stem.lower()
        data = json.loads(path.read_text())
        yield data, class_name, file_name, rel_module, type_folder


def ensure_package(path: Path):
    path.mkdir(parents=True, exist_ok=True)
    init = path / "__init__.py"
    if not init.exists():
        init.write_text("")


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--version", required=True)
    args = parser.parse_args()
    version = args.version

    url = f"https://github.com/{REPO}/releases/download/{version}/definitions.zip"
    print(f"Downloading {url}...")
    with urllib.request.urlopen(url) as response:
        payload = response.read()

    extract_dir = Path(os.environ.get("TMPDIR", "/tmp")) / f"hercules-defs-{version}"
    if extract_dir.exists():
        shutil.rmtree(extract_dir)
    with zipfile.ZipFile(io.BytesIO(payload)) as zf:
        zf.extractall(extract_dir)
    defs_dir = extract_dir / "definitions"

    # Collect data type defs first (needed to resolve references).
    data_type_defs: List[DataTypeDef] = []
    generic_key_set: Set[str] = set()
    for data, class_name, file_name, rel_module, type_folder in walk_defs(defs_dir):
        if type_folder != "data_types" or not data.get("identifier") or not data.get("type"):
            continue
        keys = [k.split("extends")[0].strip() for k in (data.get("genericKeys") or [])]
        generic_key_set.update(keys)
        data_type_defs.append(
            DataTypeDef(data["identifier"], data["type"], set(keys), rel_module, file_name)
        )
    ref_map = {d.identifier: d for d in data_type_defs}

    if OUTPUT_DIR.exists():
        shutil.rmtree(OUTPUT_DIR)
    ensure_package(OUTPUT_DIR)

    generated: List[str] = []

    def write_file(rel_module: str, file_name: str, content: str):
        segments = [sanitize_segment(s) for s in Path(rel_module).parts]
        out_dir = OUTPUT_DIR
        for seg in segments:
            out_dir = out_dir / seg
            ensure_package(out_dir)
        (out_dir / f"{file_name}.py").write_text(content)
        generated.append(".".join([*segments, file_name]))

    # Data types.
    for dt in data_type_defs:
        write_file(dt.rel_module, dt.file_name, generate_data_type(dt, ref_map, dt.generic_keys))

    # Runtime functions and flow types.
    for data, class_name, file_name, rel_module, type_folder in walk_defs(defs_dir):
        if type_folder == "runtime_flow_types":
            content = generate_runtime_flow_type(data, class_name)
        elif type_folder == "runtime_functions":
            content = generate_runtime_function(data, class_name)
        else:
            continue
        write_file(rel_module, file_name, content)

    # Top-level package imports every generated module (runs registration).
    index_lines = [
        f"import hercules.definitions.{name}  # noqa: F401"
        for name in sorted(generated)
    ]
    (OUTPUT_DIR / "__init__.py").write_text("\n".join(index_lines) + "\n")

    shutil.rmtree(extract_dir, ignore_errors=True)
    print(f"Generated {len(generated)} definition files in {OUTPUT_DIR}")


if __name__ == "__main__":
    sys.exit(main())
